x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java

```diff
@@ -245,55 +256,107 @@ public void clusterStateProcessed(ClusterState oldState, ClusterState newState)
-            throw new ElasticsearchStatusException(
-                "cannot create new assignment for model [{}] while feature reset is in progress.",
-                RestStatus.CONFLICT,
-                params.getModelId()
-            );
-        }
-        TrainedModelAssignmentMetadata.Builder builder = TrainedModelAssignmentMetadata.builder(currentState);
-        if (builder.hasModel(params.getModelId())) {
-            throw new ResourceAlreadyExistsException("assignment for model with id [{}] already exist", params.getModelId());
-        }
-        TrainedModelAssignment.Builder assignmentBuilder = TrainedModelAssignment.Builder.empty(params);
-
-        Set<String> shuttingDownNodes = nodesShuttingDown(currentState);
-        Map<String, String> nodeToReason = new TreeMap<>();
-        for (DiscoveryNode node : currentState.getNodes()) {
-            if (StartTrainedModelDeploymentAction.TaskParams.mayAssignToNode(node) && shuttingDownNodes.contains(node.getId()) == false) {
-                Optional<String> maybeError = nodeHasCapacity(currentState, params, node);
-                if (maybeError.isPresent()) {
-                    nodeToReason.put(node.getName(), maybeError.get());
-                } else {
-                    assignmentBuilder.addNewRoutingEntry(node.getId());
-                }
+        return update(currentState, rebalanceAssignments(currentState, Optional.of(params)));
+    }
+
+    private void rebalanceAssignments(
+        ClusterState clusterState,
+        Optional<StartTrainedModelDeploymentAction.TaskParams> modelToAdd,
+        String reason,
+        ActionListener<TrainedModelAssignmentMetadata> listener
+    ) {
+        threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
+            logger.debug(() -> format("Rebalancing model allocations because [%s]", reason));
+            TrainedModelAssignmentMetadata.Builder rebalancedMetadata;
+            try {
+                rebalancedMetadata = rebalanceAssignments(clusterState, modelToAdd);
+            } catch (Exception e) {
+                listener.onFailure(e);
+                return;
             }
-        }
-        if (nodeToReason.isEmpty() == false) {
-            assignmentBuilder.setReason(
-                nodeToReason.entrySet()
-                    .stream()
-                    .map(entry -> String.format(Locale.ROOT, "Not allocating on node [%s]. Reason: %s", entry.getKey(), entry.getValue()))
-                    .collect(Collectors.joining("|"))
+            submitUnbatchedTask(reason, new ClusterStateUpdateTask() {
+                @Override
+                public ClusterState execute(ClusterState currentState) {
+
+                    if (areClusterStatesCompatibleForRebalance(clusterState, currentState)) {
+                        return update(currentState, rebalancedMetadata);
+                    }
+                    rebalanceAssignments(currentState, modelToAdd, reason, listener);
```

Comment:
This is tricky. I am not sure here :/. Since it asynchronously kicks off another `rebalanceAssignments` it should be ok? 

I am just wondering how this behaves when there is a flurry of updates (maybe turning on some security integrations while creating multiple allocations? Does this lock up cluster state???)

Reply:
As this is asynchronous it shouldn't block. The method will return `currentState` and the cluster state update task will finish being a noop.

Reply:
My main concern is that a bunch of different allocations start, a flurry of callbacks keep executing and we cripple the cluster.

Reply:
Batching should alleviate this as all the tasks of the same type would be handled at the same time (and thus the number of times we do this async double check would be reduced dramatically).